r/R/dplyr-functions.R

```diff
@@ -330,6 +330,38 @@ arrow_string_join_function <- function(null_handling, null_replacement = NULL) {
   }
 }
 
+# Arrow does not supports a locale option for string case conversion functions,
+# contrast to stringr's API, so the `locale` argument is only valid for the
+# standard/default ones: "en", "C", and "POSIX". The following are string
+# functions that take a `locale` option as its second argument:
+#   str_to_lower
+#   str_to_upper
+#   str_to_title
+.valid_locales_for_string_functions <- list("en", "C", "POSIX")
+
+arrow_string_function_with_locale_arg <- function(func, string, locale) {
+  assert_that(
+    locale %in% .valid_locales_for_string_functions,
+    msg = paste(
+      "`locale` must be any of: ",
+      paste(.valid_locales_for_string_functions, collapse=",")
+    )
+  )
+  Expression$create(func, string)
+}
+
+nse_funcs$str_to_lower <- function(string, locale = "en") {
+  arrow_string_function_with_locale_arg("utf8_lower", string, locale)
```

Comment:
I would ditch the function above and rewrite these like this:

```suggestion
  if (!identical(locale, "en")) {
    stop("Providing 'locale' to 'str_to_lower' not supported in Arrow; to change locale, use 'Sys.setlocale()'", call. = FALSE)
  }
  Expression$create("utf8_lower", string)
```

Thoughts @jonkeane @thisisnic ? 

(I'd also leave a comment referencing the jira issue that would support passing locale to the string kernels)

Reply:
FYI, when an invalid `locale` is provided the following message is triggered in a (Table -> mutate -> collect) pipeline:
```r
Warning: In str_to_title(x, locale = "es"), Providing 'locale' to 'str_to_title' is not supported in Arrow; to change locale use 'Sys.setlocale()'; pulling data into R
Error: Problem with `mutate()` column `r`.
ℹ `r = str_to_title(x, locale = "es")`.
✖ could not find function "str_to_title"
Run `rlang::last_error()` to see where the error occurred.
```